pinot-controller/src/main/java/org/apache/pinot/controller/util/ConsumingSegmentInfoReader.java

```diff
@@ -131,6 +135,51 @@ private String generateServerURL(String tableNameWithType, String endpoint) {
     return String.format("%s/tables/%s/consumingSegmentsInfo", endpoint, tableNameWithType);
   }
 
+  /**
+   * Utility method to derive ingestion status from consuming segment Info. Status is HEALTHY if
+   * consuming segment info specifies CONSUMING state for all active segments across all servers
+   * including replicas.
+   */
+  public TableStatus.IngestionStatus getIngestionStatus(String tableNameWithType, int timeoutMs) {
+    try {
+      ConsumingSegmentsInfoMap consumingSegmentsInfoMap = getConsumingSegmentsInfo(tableNameWithType, timeoutMs);
+      for (Map.Entry<String, List<ConsumingSegmentInfo>> consumingSegmentInfoEntry : consumingSegmentsInfoMap._segmentToConsumingInfoMap
+          .entrySet()) {
+        String segmentName = consumingSegmentInfoEntry.getKey();
+        List<ConsumingSegmentInfo> consumingSegmentInfoList = consumingSegmentInfoEntry.getValue();
+        if (consumingSegmentInfoList == null || consumingSegmentInfoList.isEmpty()) {
+          String errorMessage = "Did not get any response from servers for segment: " + segmentName;
+          return TableStatus.IngestionStatus.newIngestionStatus(TableStatus.IngestionState.UNHEALTHY, errorMessage);
+        }
+
+        // Check if any responses are missing
+        Set<String> serversForSegment = _pinotHelixResourceManager.getServersForSegment(tableNameWithType, segmentName);
+        if (serversForSegment.size() != consumingSegmentInfoList.size()) {
+          Set<String> serversResponded =
+              consumingSegmentInfoList.stream().map(c -> c._serverName).collect(Collectors.toSet());
+          serversForSegment.removeAll(serversResponded);
+          String errorMessage =
+              "Not all servers responded for segment: " + segmentName + " Missing servers : " + serversForSegment;
+          return TableStatus.IngestionStatus.newIngestionStatus(TableStatus.IngestionState.UNHEALTHY, errorMessage);
+        }
+
+        for (ConsumingSegmentInfo consumingSegmentInfo : consumingSegmentInfoList) {
+          if (consumingSegmentInfo._consumerState
+              .equals(RealtimeSegmentDataManager.ConsumerState.NOT_CONSUMING.toString())) {
```

Comment:
The controller should not depend on the server. RealtimeSegmentDataManager belongs in pinot-sever right? If it is in core now, it should probably be moved to pinot-server anyway.

Reply:
Good point. Refactored ConsumerState to be in CommonConstants instead.